python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -0,0 +1,197 @@
+from typing import Any, Callable, Optional, Dict
+
+from transformers.trainer import Trainer
+from torch.utils.data import Dataset as TorchDataset
+
+from ray.train.torch import TorchConfig
+from ray.ml.trainer import GenDataset
+from ray.ml.train.integrations.torch import TorchTrainer
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.util import PublicAPI
+from ray.ml.constants import TRAIN_DATASET_KEY, EVALUATION_DATASET_KEY
+
+
+@PublicAPI(stability="alpha")
+class HuggingFaceTrainer(TorchTrainer):
+    """A Trainer for data parallel HuggingFace Transformers on PyTorch training.
+
+    This Trainer runs the ``transformers.Trainer.train()`` method on multiple
+    Ray Actors. The training is carried out in a distributed fashion through PyTorch
+    DDP. These actors already have the necessary torch process group already
+    configured for distributed PyTorch training.
+
+    The training function ran on every Actor will first run the
+    specified ``trainer_init_per_worker`` function to obtain an instantiated
+    ``transformers.Trainer`` object. The ``trainer_init_per_worker`` function
+    will have access to preprocessed train and evaluation datsets.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards, with each Actor training on a single shard.
+    All the other datasets will not be split.
+
+    Please note that if you use a custom ``transformers.Trainer`` subclass,
+    the ``get_train_dataloader`` method will be overriden to disable distributed
+    sampling, as the dataset will already be sharded.
+
+    Hugging Face loggers will be automatically disabled, and the ``local_rank``
+    argument in ``TrainingArguments`` will be automatically set.
+
+    Example:
+        .. code-block:: python
+            # Based on
+            # huggingface/notebooks/examples/language_modeling_from_scratch.ipynb
```

Comment:
This example would be great to add in the examples dir, but is there a simpler one we can use for the docstring?

Reply:
The dataset and model need to be in exact formats for the transformers library to work, so using dummy data is not really possible here. We can try to simplify it later, perhaps.